src/linear_algebra/contraction.lean

```diff
@@ -107,6 +117,52 @@ equivalence. -/
 @[simp] noncomputable def dual_tensor_hom_equiv : (module.dual R M) ⊗[R] N ≃ₗ[R] M →ₗ[R] N :=
 dual_tensor_hom_equiv_of_basis (module.free.choose_basis R M)
 
+variables [module.free R N] [module.finite R N]
+
+open module
+
+variables [module.finite R (M ⊗[R] N)] --Temporary until this is made an instance by #13705
+
+/--
+When `M` and `N` are free `R` modules, the map `hom_tensor_hom_map` is an equivalence. Note that
+`hom_tensor_hom_equiv` is not defined directly in terms of `hom_tensor_hom_map`, but the equivalence
+between the two is given by `hom_tensor_hom_equiv_to_linear_map` and `hom_tensor_hom_equiv_apply`.
+-/
+noncomputable
+def hom_tensor_hom_equiv : (M →ₗ[R] P) ⊗[R] (N →ₗ[R] Q) ≃ₗ[R] (M ⊗[R] N →ₗ[R] P ⊗[R] Q) :=
+  congr (dual_tensor_hom_equiv R M P).symm (dual_tensor_hom_equiv R N Q).symm ≪≫ₗ
+  tensor_tensor_tensor_comm R (dual R M) P (dual R N) Q ≪≫ₗ
+  tensor_product.congr dual_tensor_dual_equiv (linear_equiv.refl R (P ⊗ Q)) ≪≫ₗ
+  dual_tensor_hom_equiv R (M ⊗ N) (P ⊗ Q)
```

Comment:
You could introduce auxiliary definitions
```
noncomputable def ltensor_hom_equiv_hom_ltensor : P ⊗[R] (M →ₗ[R] Q) ≃ₗ[R] (M →ₗ[R] P ⊗[R] Q) :=
congr (linear_equiv.refl R P) (dual_tensor_hom_equiv R M Q).symm ≪≫ₗ
  tensor_product.left_comm R P _ Q ≪≫ₗ dual_tensor_hom_equiv R M _

noncomputable def rtensor_hom_equiv_hom_rtensor : (M →ₗ[R] P) ⊗[R] Q ≃ₗ[R] (M →ₗ[R] P ⊗[R] Q) :=
congr (dual_tensor_hom_equiv R M P).symm (linear_equiv.refl R Q) ≪≫ₗ
  tensor_product.assoc R _ P Q ≪≫ₗ dual_tensor_hom_equiv R M _
```
then define this as
```
noncomputable
def hom_tensor_hom_equiv : (M →ₗ[R] P) ⊗[R] (N →ₗ[R] Q) ≃ₗ[R] (M ⊗[R] N →ₗ[R] P ⊗[R] Q) :=
rtensor_hom_equiv_hom_rtensor R M P _ ≪≫ₗ
  (linear_equiv.refl R M).arrow_congr (ltensor_hom_equiv_hom_ltensor R N _ Q) ≪≫ₗ
  lift.equiv R M N _
```
which avoids the need to assume `variables [module.finite R (M ⊗[R] N)]`

Reply:
The ```variables [module.finite R (M ⊗[R] N)]``` won't be a problem anymore once #13733 gets merged though.

Reply:
Yes, but I think these tensor-hom commutation equivalences have independent interest, and you'd also prove lemmas about how they act on simple tensors, which you can use in the proof of `hom_tensor_hom_equiv_to_linear_map`, essentially breaking the proof down to smaller pieces.

Reply:
Fair enough

Reply:
The forward map can be constructed unconditionally:
```
def ltensor_hom_to_hom_ltensor : P ⊗[R] (M →ₗ[R] Q) →ₗ[R] (M →ₗ[R] P ⊗[R] Q) :=
tensor_product.lift (llcomp R M Q _ ∘ₗ mk R P Q)
```
and you might try to show
```
lemma ltensor_hom_equiv_hom_ltensor_to_linear_map :
  (ltensor_hom_equiv_hom_ltensor R M P Q).to_linear_map = ltensor_hom_to_hom_ltensor R M P Q :=
sorry
```
I'm not yet familiar with the way to do these proofs, and I'll continue tomorrow.

Reply:
I've done the changes you propose!

Reply:
Thanks! I'll take a look and see whether I can shorten the proofs of `_to_linear_map` lemmas.

Reply:
They're pretty long because I used `simp only` instead of `simp`, otherwise they would be very short. I'm not sure what are mathlib's conventions on this matter, but I used `simp_only` because it reduces the loading time significantly